content/post/reliable-change-detection.md

```diff
@@ -0,0 +1,258 @@
+No matter which solution they pick, they must deal with frustrating and hard-to-debug limitations.
+
+### Solution 5: Per-system change detection
+
+Fed up with this mess, we store one copy of the relevant change detection data in each of our systems, updating it whenever the system is changed.
+In effect, this behaves much like event channels, with change events getting sent out to the appropriate systems based on the data they require access to.
+Under the hood, this could be done quite easily using a [system-local `Local` resource](https://docs.rs/bevy/0.5.0/bevy/ecs/system/struct.Local.html).
+
+This is ergonomic (1), automatic (2), and perfectly reliable, as we only update this data when the change has been seen.
+And, fascinatingly, this is the first solution that handles both skipped and repeating systems properly, as we're finally able to desynchronize the `Changed` status of each component by system and persist changes for more than a frame or two.
+
+However, this is a *large* amount of data duplication, even if we're clever and only store it on the systems that require `Changed<T>` data.
+This hurts our performance in unacceptable ways: particularly the memory usage (criteria 3).
+
+## A no-compromise solution: Ring-buffered change-ticks
+
+Despite these frustrations, we seem to be successfully stumbling towards a solution.
+We've learned that any successful solution must:
+
+1. Persist changes for more than a frame or two to avoid missing systems that are skipped.
+2. Store *some* data on a per-system basis to avoid double counting, especially when working with repeated systems.
+3. Report changes immediately (or at least, before anyone else is allowed to look at those changes).
+4. Avoid naively duplicating all of the change detection data between each of our systems.
+
+Avoiding duplication is where the real magic begins.
+
+To do so, we store timers in three separate locations:
+
+* On the `World` as a global indicator of current time. Whenever a system begins execution, [this increments once](https://github.com/bevyengine/bevy/pull/1471/files#diff-5499ffc12ee4011981ff6d5069a1986197d824648c65ed24056e6b2772b3cd81R146).
+* On each piece of `Resource` or `Component` data, [marking when the data was last changed](https://github.com/bevyengine/bevy/pull/1471/files#diff-0e94997025571f709abd7cac97f03bb4e8ec8bf29650068a7e5ec07170011892R137) in terms of the global timer.
+* On each `System`, marking when the system last ran in terms of the global timer.
+
+If the data's change tick is greater than the system's change tick, the data has changed since it was last processed.
+We use [wrapping subtraction]((https://github.com/bevyengine/bevy/pull/1471/files#diff-0e94997025571f709abd7cac97f03bb4e8ec8bf29650068a7e5ec07170011892R137) to determine what "greater than" means, giving us **ring buffer** behavior, allowing our change detection to continue functioning indefinitely rather than running out of space as time goes on.
```

Comment:
Typo:
```suggestion
We use [wrapping subtraction](https://github.com/bevyengine/bevy/pull/1471/files#diff-0e94997025571f709abd7cac97f03bb4e8ec8bf29650068a7e5ec07170011892R137) to determine what "greater than" means, giving us **ring buffer** behavior, allowing our change detection to continue functioning indefinitely rather than running out of space as time goes on.
```